### PR TITLE
Leaves-- Janice

### PR DESCRIPTION
**Pull request by jaitch**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a top_ten method for the Work model, and a helper method for that method, sort_works. As their names suggest, they were the methods that sorted works by descending vote number.
Describe how you approached testing that model method. What edge cases did you come up with? | I tested them with fewer than 10 works for the top_ten method, some zero vote works.
What are `session` and `flash`? What is the difference between them? | `session` and `flash` are hash-like objects that store data. `session` stores data (user data, in our case) for the duration of the session. `flash` stores it until the redirect or for an even shorter amount of time, when it's `flash.now`.
What was one thing that you gained more clarity on through this assignment? | I became more comfortable with manipulating data and using the relationships between models. 
What is the Heroku URL of your deployed application? | https://jan-mediaranker-app.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | It was an interesting challenge to replicate an existing site, but this made it difficult to explore Bootstrap--it was hard to figure out what was Bootstrap styling and what (if any) wasn't. In terms of back-end work, I appreciate that there were a few, but not too many and not too earth-shattering, tools to try out with this project. A lot of Rails feels uncomfortable until we practice, so it was nice to keep practicing without a hugely new, difficult leap (while still feeling improvement). I appreciated the opportunity to do this project by myself, as the RideShare project was shared between partners.

### CODE REVIEW

File: app/views/homepages/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+<section class="top-ten__container">
+  <section class="top-ten__list-container">
+  <h2 class="top-ten__header"> Top Books: </h2>
+<ul>
+<% chosen = @works.where(category: "book") %>
+<% chosen = Work.top_ten(chosen) %>
+<% chosen.each do |book| %>
+<li> <%= link_to book.title, work_path(book) %>
+by
+<%= book.creator %>
```
Comment (beccaelenzil) [inline app/views/homepages/index.html.erb:28]: A partial view would be good to utilize here for this repeated code.

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
+class UsersController < ApplicationController
+  def index
+    @users = User.all
+  end
+
+  def show
+    @user = User.find_by(id: params[:id])
```
Comment (beccaelenzil) [inline app/controllers/users_controller.rb:7]: This is good functionality to put in a controller filter method in the Applications Controller.

**Comment by beccaelenzil** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | yes
Optional: Look and feel is similar to the original (consider styling and layout) | yes
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | yes
Can go into a work's show page | yes
Verify unable to vote on a work, and get a flash message | yes
Can edit this work successfully, and get a flash message | yes
Can go to "View all media" page and see three lists of works, sorted by vote | yes
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | yes
Can create a new work successfully. Note the URL for this work's show page | yes
Can delete this work successfully |yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | yes
Verify that the "View all users" page lists no users (except what is produced by db seeding) |
**Log in** |
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | yes
Your username is listed in "View all users" page | yes
Verify that number of votes determines the Media Spotlight | yes
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page |
Voting on the same work twice produces an error/flash message, and there is no extra vote | yes
**Log out** |
Logging out showed a flash message and changed the UI | yes
Logging in as a new user creates a new user | yes
Logging in as an already existing user has a specific flash message | yes

### Targeted Code Review

Area | yes / no
--- | ---
git commits were small and atomic, with useful messages | yes
Routes file uses `resources` for works |yes
Routes file shows intention in limiting routes for voting, log-in functionality, and users | yes
The homepage view, all media view, and new works view use semantic HTML | yes
The homepage view, all media view, and new works view use partials when appropriate | see comment
The model for media (likely named `work.rb`) `has_many` votes | yes
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | yes
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | no, see comment
Some controller, likely the `WorksController`, has a controller filter for finding a work | yes
The `WorksController` uses strong params | yes
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | yes

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | yes
User model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | yes
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Work model has tests with a section on all business logic methods in the model, including their edge cases | yes

## Overall Feedback
Great work overall! Your code is well written and well tested. It is clear that the learning goals for this project were met. Keep up the hard work! I've left a couple in line comment for you to review.
